### PR TITLE
Pod Idler

### PR DESCRIPTION
**Pull request by alexeykazakov**
This PR introduces `Idler` controller which watches cluster scoped `Idler` resources as the primary resource.
`Idler` resource will be later (as soon as this PR is merged and deployed to the member cluster) added to the Tier yamls.

The flow is the following:
- Idler resources are added to Tier's cluster resource yamls. See https://github.com/codeready-toolchain/host-operator/pull/245
- NSTemplateSet controller creates an `Idler` resource per namespace during user provisioning. One Idler (cluster scoped) manages one namespace (Idler name == namespace name).
- Idler controller watches Idlers (primary resource) and pods (secondary resources). All pods from the namespace managed by the Idler are tracked by the Idler and if they are running for too long then the pod is deleted directly (if a standalone pod) or by scaling down the corresponding controller (Deployment, ReplicaSet, etc).

This PR doesn't have any e2e tests because it requires https://github.com/codeready-toolchain/host-operator/pull/245 to be merged first to enable the idling.
Here is how we can properly test it:
- Review and merge this PR. Idler controller is now in member cluster but it's not enabled because there is no Idler resources in the cluster.
- Add e2e tests to https://github.com/codeready-toolchain/host-operator/pull/245. Now the paired e2e PR has Idling enabled and can do the e2e testing.
- Review and merge https://github.com/codeready-toolchain/host-operator/pull/245.

### CODE REVIEW

File: pkg/controller/idler/idler_controller.go [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+		return err
+	}
+	newStatusPods := make([]toolchainv1alpha1.Pod, 0, 10)
+	for _, pod := range podList.Items {
+		podLogger := logger.WithValues("Pod.Name", pod.Name)
+		podLogger.Info("Pod", "Pod.Phase", pod.Status.Phase)
+		if trackedPod := findPodByName(idler, pod.Name); trackedPod != nil {
+			// Already tracking this pod. Check the timeout.
+			newStatusPods = append(newStatusPods, *trackedPod) // keep tracking
+			if time.Now().After(trackedPod.StartTime.Add(time.Duration(idler.Spec.TimeoutSeconds) * time.Second)) {
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/idler/idler_controller.go:131]: It's a good point. I think we can just ignore Idlers with 0 timeout. Will add it.

File: pkg/controller/idler/idler_controller.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+}
+
+// scaleControllerToZero checks if the object has an owner controller (Deployment, ReplicaSet, etc)
+// and scales the owner down to zero and returns "true".
+// Otherwise returns "false".
+func (r *ReconcileIdler) scaleControllerToZero(logger logr.Logger, meta metav1.ObjectMeta) (bool, error) {
+	owners := meta.GetOwnerReferences()
+	for _, owner := range owners {
+		if owner.Controller != nil && *owner.Controller {
+			switch owner.Kind {
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/idler/idler_controller.go:164]: I was thinking about a job too but thought that a job is for one time action. Didn't think about the restartPolicy thing. Good point. I will add Job to the list of supported controllers too.

File: pkg/controller/idler/idler_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+
+				t.Run("Third Reconcile. Stop tracking deleted pods.", func(t *testing.T) {
+					//when
+					res, err := reconciler.Reconcile(req)
+
+					// then
+					require.NoError(t, err)
+					// Tracking existing pods only.
+					memberoperatortest.AssertThatIdler(t, idler.Name, cl).
+						TracksPods(append(append(podsTooEarlyToKill.controlledPods, podsTooEarlyToKill.controlledPods...), podsTooEarlyToKill.standalonePods...)).
```
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/idler/idler_controller_test.go]: Yes it is! Fixed in https://github.com/codeready-toolchain/member-operator/pull/179/commits/07986a98ee7414eb047fc497a435106992b17c1a

File: pkg/controller/idler/idler_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/idler/idler_controller.go:137]: Done in https://github.com/codeready-toolchain/member-operator/pull/179/commits/8d77bba102a84bb14e00842f85b5b8290bdc6c30

File: pkg/controller/idler/idler_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/controller/idler/idler_controller.go:171]: Done in https://github.com/codeready-toolchain/member-operator/pull/179/commits/52ed5ded1d09af2c9e7c92d8b898f5c553d54687

File: pkg/controller/idler/idler_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+			res, err := reconciler.Reconcile(req)
+
+			// then
+			require.NoError(t, err)
+			assert.Equal(t, reconcile.Result{}, res)
+			memberoperatortest.AssertThatIdler(t, idler.Name, cl).HasConditions(memberoperatortest.FailedToIdle("timoutSeconds should be bigger than 0"))
+		}
+
+		assertInvalidTimout(0)
+		assertInvalidTimout(-1)
```
Comment (rajivnathan) [inline pkg/controller/idler/idler_controller_test.go:270]:
```suggestion
	t.Run("Fail if Idler.Spec.TimeoutSec is invalid", func(t *testing.T) {
		assertInvalidTimeout := func(timeout int32) {
			// given
			idler := &v1alpha1.Idler{
				ObjectMeta: metav1.ObjectMeta{
					Name: "john-dev",
				},
				Spec: v1alpha1.IdlerSpec{TimeoutSeconds: timeout},
			}
			reconciler, req, cl := prepareReconcile(t, idler.Name, idler)

			// when
			res, err := reconciler.Reconcile(req)

			// then
			require.NoError(t, err)
			assert.Equal(t, reconcile.Result{}, res)
			memberoperatortest.AssertThatIdler(t, idler.Name, cl).HasConditions(memberoperatortest.FailedToIdle("timeoutSeconds should be bigger than 0"))
		}

		assertInvalidTimeout(0)
		assertInvalidTimeout(-1)
```
typos

Comment (alexeykazakov) [reply pkg/controller/idler/idler_controller_test.go:270]: I'm a terrible speller :)
Thank you!

**Comment by alexeykazakov** [comment]
@rajivnathan thank you for the review!

**Pull request merged**